src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -376,6 +376,227 @@ bool LibraryCallKit::inline_vector_nary_operation(int n) {
+    }
+    return false; // not supported
+  }
+
+  Node* opd1 = NULL; Node* opd2 = NULL; Node* opd3 = NULL;
+  switch (n) {
+    case 3: {
+      opd3 = unbox_vector(argument(7), vbox_type, elem_bt, num_elem);
+      if (opd3 == NULL) {
+        if (C->print_intrinsics()) {
+          tty->print_cr("  ** unbox failed v3=%s",
+                        NodeClassNames[argument(7)->Opcode()]);
+        }
+        return false;
+      }
+      // fall-through
+    }
+    case 2: {
+      opd2 = unbox_vector(argument(6), vbox_type, elem_bt, num_elem);
+      if (opd2 == NULL) {
+        if (C->print_intrinsics()) {
+          tty->print_cr("  ** unbox failed v2=%s",
+                        NodeClassNames[argument(6)->Opcode()]);
+        }
+        return false;
+      }
+      // fall-through
+    }
+    case 1: {
+      opd1 = unbox_vector(argument(5), vbox_type, elem_bt, num_elem);
+      if (opd1 == NULL) {
+        if (C->print_intrinsics()) {
+          tty->print_cr("  ** unbox failed v1=%s",
+                        NodeClassNames[argument(5)->Opcode()]);
+        }
+        return false;
+      }
+      break;
+    }
+    default: fatal("unsupported arity: %d", n);
+  }
+
+  Node* mask = NULL;
+  if (use_mask) {
+    ciKlass* mbox_klass = mask_klass->const_oop()->as_instance()->java_lang_Class_klass();
+    assert(is_vector_mask(mbox_klass), "argument(2) should be a mask class");
+    const TypeInstPtr* mbox_type = TypeInstPtr::make_exact(TypePtr::NotNull, mbox_klass);
+    mask = unbox_vector(argument(n + 5), mbox_type, elem_bt, num_elem);
+    if (mask == NULL) {
+        if (C->print_intrinsics()) {
+          tty->print_cr("  ** unbox failed mask=%s",
+                        NodeClassNames[argument(n + 5)->Opcode()]);
+        }
+        return false;
+      }
+  }
+
+  Node* operation = NULL;
+  if (opc == Op_CallLeafVector) {
+    assert(UseVectorStubs, "sanity");
+    operation = gen_call_to_svml(opr->get_con(), elem_bt, num_elem, opd1, opd2);
+    if (operation == NULL) {
+      if (C->print_intrinsics()) {
+        tty->print_cr("  ** svml call failed");
+      }
+      return false;
+     }
+  } else {
+    const TypeVect* vt = TypeVect::make(elem_bt, num_elem);
+    switch (n) {
+      case 1:
+      case 2: {
+        operation = VectorNode::make(sopc, opd1, opd2, vt);
+        break;
+      }
+      case 3: {
+        operation = VectorNode::make(sopc, opd1, opd2, opd3, vt);
+        break;
+      }
+      default: fatal("unsupported arity: %d", n);
+    }
+  }
+
+  // Currently just focus on binary mask operation.
+  if (n == 2 && mask != NULL) {
+    if (sopc != 0 &&
+        Matcher::match_rule_supported_masked_vector(sopc, num_elem, elem_bt) &&
+        Matcher::match_rule_supported(Op_VectorToMask)) {
+      const TypeVMask* vmask_type = TypeVMask::make(elem_bt, num_elem);
+      mask = gvn().transform(new VectorToMaskNode(mask, vmask_type));
+      operation->add_req(mask);
```

Comment:
Following is the link to a reference implementation (PoC) for X86: 
http://cr.openjdk.java.net/~jbhateja/avx512_masked_operation_optimization/webrev.02/src/hotspot/share/opto/vectornode.cpp.udiff.html
This demonstrates creation of new masked operation node which folds blend+vector operation graph patterns, there is one new Ideal node for each kind of operation i.e. ternary, binary and unary. New node also carry the meta data info which specifies the kind of masked operation being performed.